Zend/zend_inheritance.c

```diff
@@ -2150,20 +2128,18 @@ typedef struct _zend_abstract_info {
 
 static void zend_verify_abstract_class_function(zend_function *fn, zend_abstract_info *ai) /* {{{ */
 {
-	if (fn->common.fn_flags & ZEND_ACC_ABSTRACT) {
-		if (ai->cnt < MAX_ABSTRACT_INFO_CNT) {
-			ai->afn[ai->cnt] = fn;
-		}
-		if (fn->common.fn_flags & ZEND_ACC_CTOR) {
-			if (!ai->ctor) {
-				ai->cnt++;
-				ai->ctor = 1;
-			} else {
-				ai->afn[ai->cnt] = NULL;
-			}
-		} else {
+	if (ai->cnt < MAX_ABSTRACT_INFO_CNT) {
+		ai->afn[ai->cnt] = fn;
+	}
+	if (fn->common.fn_flags & ZEND_ACC_CTOR) {
```

Comment:
The way this affects ReflectionMethod->isConstructor should be documented in `UPGRADING` (e.g. for `public function __construct()` in an interface), and https://www.php.net/manual/en/reflectionmethod.isconstructor.php .

In php <= 7.4, it returned false for abstract constructors. I'd still think this change of returning true is more correct.
In php 8.0, it now returns true.

```
    protected function canMockMethod(ReflectionMethod $method)
    {
        if ($method->isConstructor() ||
            $method->isFinal() ||
            $method->isPrivate() ||
            isset($this->blacklistedMethodNames[$method->getName()])) {
            return false;
        }

        return true;
```

Reply:
Uh, how is this related to this patch? This code is about generating error messages.

Reply:
Sorry, I was in a hurry and misread what this was doing before reading the thread - it's reading ZEND_ACC_CTOR, but not setting it